File: app/src/main/java/com/example/pc/oreo/WifiCommand.java
package com.example.pc.oreo;

import org.json.JSONArray;
import org.json.JSONException;

public class WifiCommand {
    private byte[] rawData;
    private WifiCommandCode type;
    private byte[] payload;
    private int[] flightDataPayload;

    public WifiCommand(byte[] rawData) {
        this.rawData = rawData;
        parseData();
    }

    public WifiCommand(int type) {
        for (WifiCommandCode code : WifiCommandCode.values()) {
            if (type == code.getValue()) {
                this.type = code;
                break;
            }
        }
    }

    private void parseData() {
        for (WifiCommandCode code : WifiCommandCode.values()) {
            if (rawData[1] == code.getValue()) {
                type = code;
                break;
            }
        }
        if (type == WifiCommandCode.FLIGHT_DATA) {
            int payloadLength = rawData.length - 2;
            payload = new byte[payloadLength];
            System.arraycopy(rawData, 2, payload, 0, payloadLength);
            try {
                JSONArray arr = new JSONArray(new String(payload));
                flightDataPayload = new int[arr.length()];
                for (int i = 0; i < arr.length(); i++) {
                    flightDataPayload[i] = arr.optInt(i);
                }
            } catch (JSONException e) {
                flightDataPayload = new int[0];
            }
        } else {
            int payloadLength = type.getLength() - 2;
            payload = new byte[payloadLength];
            System.arraycopy(rawData, 2, payload, 0, payloadLength);
        }
    }

    public WifiCommandCode getType() {
        return type;
    }

    public int[] getFlightDataPayload() {
        return flightDataPayload;
    }

    public enum WifiCommandCode {
        REQUEST_CONNECTION(100, 2),
        APPROVE_CONNECTION(101, 2),
        JOYSTICK_CONTROL(102, 14),
        TAKE_OFF(103, 2),
        LAND(104, 2),
        REQUEST_STREAM(105, 2),
        FRAME_DATA(106, 2),
        FLIGHT_DATA(107, 36),
        CAMERA_SETTINGS(108, 10);

        private final byte value;
        private final byte length;

        WifiCommandCode(int value, int length) {
            this.value = (byte) value;
            this.length = (byte) length;
        }

        public byte getValue() {
            return value;
        }

        public byte getLength() {
            return length;
        }
    }
}

File: app/src/main/java/com/example/pc/oreo/WifiCommand.kt
package com.example.pc.oreo

import org.json.JSONArray
import org.json.JSONException

class WifiCommand {
    private var rawData: ByteArray? = null
    var type: WifiCommandCode? = null
        private set
    private lateinit var payload: ByteArray
    lateinit var driveDataPayload: IntArray

    constructor(rawData: ByteArray) {
        this.rawData = rawData
        parseData()
    }

    constructor(type: Int) {
        this.type = WifiCommandCode.values().first { code -> code.value.toInt() == type }
    }

    private fun parseData() {
        type = WifiCommandCode.values().first { code -> code.value == rawData?.first() }
        when (type) {
            WifiCommandCode.DRIVE_DATA -> {
                val payloadLength = rawData!!.size - 2
                payload = ByteArray(payloadLength)
                System.arraycopy(rawData, 2, payload, 0, payloadLength)
                try {
                    val arr = JSONArray(String(payload))
                    driveDataPayload = IntArray(arr.length())
                    for (i in 0 until arr.length()) {
                        driveDataPayload[i] = arr.optInt(i)
                    }
                } catch (e: JSONException) {
                    driveDataPayload = IntArray(0)
                }
            }
            else -> {
                val payloadLength = type!!.length - 2
                payload = ByteArray(payloadLength)
                System.arraycopy(rawData, 2, payload, 0, payloadLength)
            }
        }
    }

    enum class WifiCommandCode constructor(value: Int, length: Int) {
        REQUEST_CONNECTION(100, 2),
        APPROVE_CONNECTION(101, 2),
        JOYSTICK_CONTROL(102, 14),
        TAKE_OFF(103, 2),
        LAND(104, 2),
        REQUEST_STREAM(105, 2),
        FRAME_DATA(106, 2),
        DRIVE_DATA(107, 36),
        CAMERA_SETTINGS(108, 10);

        val value: Byte = value.toByte()
        val length: Byte = length.toByte()

    }
}
